src/language-js/printer-estree.js

```diff
@@ -3878,35 +3878,40 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
     ? printFunctionTypeParameters(path, options, print)
     : "";
 
-  let printed = [];
+  let parts = [];
   if (fun[paramsField]) {
-    const lastArgIndex = fun[paramsField].length - 1;
-
-    printed = path.map((childPath, index) => {
-      const parts = [];
-      const param = childPath.getValue();
+    parts = path.map((childPath) => {
+      return { value: childPath.getValue(), param: print(childPath) };
+    }, paramsField);
+  }
 
-      parts.push(print(childPath));
+  if (fun.this) {
+    fun.this.name = "this";
```

Comment:
Instead of modify the ast we can check parent in `FunctionTypeParam`

Reply:
the case for `FunctionTypeParam` prints a whole identifier, so re-implementing all the spacing logic there seems redundant. Assigning `this` to the ast gets us all this behavior for free.

Reply:
We are trying not  to [mutate the AST](https://github.com/prettier/prettier/issues/8122), abd that part seems simple

```js
node.name ?
path.call(print, "name") :
parent.this =node ? "this ":""
```

Should work

Reply:
Yea the issue was that doing this alone causes the test case above with the multi-line union types to be different between `this` params and regular params. Adding an extra condition to the printing for union types seems to fix it though.